rfcs/0084-input-aware-fetchers.md

```diff
@@ -0,0 +1,147 @@
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.10";
+  sha256 = "0000000000000000000000000000000000000000000000000000";
+}
+```
+
+Building this we get a hash mismatch as expected:
+
+```
+hash mismatch in fixed-output derivation '/nix/store/5d3k20pzgjyccmpqfina1cvbl28zxz6a-source':
+  wanted: sha256:0000000000000000000000000000000000000000000000000000
+  got:    sha256:1cx9yv62rylfv8p09pidsmqy8qim1bbjaa8pj1j8xj7vkrm0dri1
+```
+
+Now we can copy the hash into the expression:
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.10";
+  sha256 = "1cx9yv62rylfv8p09pidsmqy8qim1bbjaa8pj1j8xj7vkrm0dri1";
+}
+```
+
+And build it successfully to get the correct result in the output path
+```
+/nix/store/5d3k20pzgjyccmpqfina1cvbl28zxz6a-source
+```
+
+It is however very easy to forget to update the hash to something invalid, which as many know has been a source of much confusion, questions and frustration.
+
+However it doesn't have to be this way!
+
+# Detailed design
+[design]: #detailed-design
+
+This problem can be solved relatively easily by realizing that Nix not only uses the output hash of derivations to check whether they need to be built, but the _full_ store path, which notably also includes the _derivation name_. Thus, by ensuring that the derivation name changes if (and only if) relevant inputs change, we can force Nix to rebuild a derivation even if the output hash stays the same, therefore fixing above problem.
+
+This RFC therefore proposes to use a hash of all relevant derivation inputs as the default name of fixed-output derivations. The hash should be computed using `sha256`, and encoded using the [url-safe base64](https://tools.ietf.org/html/rfc4648#section-5), without any leading `=`, leading to a character count of 43. For reference, this hash can be computed in `nix-shell -p openssl` with
+```
+[nix-shell:~]$ echo -n 'example string' | openssl dgst -sha256 -binary | openssl base64 -A | cut -b1-42 | tr +/ -_
+rt-5KzBTohoRT08wGgKjxq1d_1BNEk3CzuYRdiPuxw
+```
+
+The following describes exactly which string should be used as the hash input for each of the fetcher types. This is necessary since there are multiple ways to fetch the same resources, e.g. with the builtin `fetchTarball` and nixpkgs' `fetchzip`. In order for them to be able to reuse the same store path, they should use the same derivation name.
+
+| Fetchers | Input string |
+| --- | --- |
+| `pkgs.fetchurl`, `<nix/fetchurl.nix>`, `builtins.fetchurl`, `nix-prefetch-url` | "fetchurl-" + first URL |
+| `pkgs.fetchzip`, `<nix/fetchurl.nix>` (with `unpack = true`), `builtins.fetchTarball`, `nix-prefetch-url --unpack` | "fetchurl-unpack-" + URL |
+| `pkgs.fetchgit`, `builtins.fetchGit` and co. | "fetchgit-" + url + "-" + rev |
```

Comment:
`pkgs.fetchgit` would also need to have `deepClone`, `leaveDotGit` and `fetchSubmodules` included, wouldn't it?

Reply:
Oh yeah, though we'll have to make sure that it harmonizes with `builtins.fetchGit` which doesn't support `deepClone` and `leaveDotGit`. So it would probably have to be something like
```nix
"fetchgit-${optionalString deepClone "deepclone-"}${optionalString leaveDotGit "leavedotgit-"}${optionalString fetchSubmodules "fetchsubmodules-"}..."
```